pkg/controller/registry/resolver/resolver_test.go

```diff
@@ -1198,3 +1198,41 @@ func TestSolveOperators_WithoutDeprecated(t *testing.T) {
 	}
 	require.EqualValues(t, expected, operators)
 }
+
+// Scenario: An operator bundle provides and requires the same API
+// Expected behavior: The resolver picks the bundle to fulfill the API requirement. One operator can satisfy its own dependency.
+func TestSolveOperators_FulfillsOwnRequirements(t *testing.T) {
+	APISet := APISet{opregistry.APIKey{"g", "v", "k", "ks"}: struct{}{}}
+
+	namespace := "olm"
+	channel := "alpha"
+	catalog := registry.CatalogKey{Name: "community", Namespace: namespace}
+
+	newSub := newSub(namespace, "packageA", channel, catalog)
+	subs := []*v1alpha1.Subscription{newSub}
+
+	fakeNamespacedOperatorCache := NamespacedOperatorCache{
+		snapshots: map[registry.CatalogKey]*CatalogSnapshot{
+			catalog: {
+				operators: []*Operator{
+					// Default channel is stable in this case
+					genOperator("packageA.v0.0.2", "0.0.2", "", "packageA", channel, catalog.Name, catalog.Namespace, APISet, APISet, nil, channel, false),
```

Comment:
I wouldn't think you'd list the API as both provided and required, since if it's provided, there's no need for the API to be resolved as a dependency. This is an interesting case though, and to really make sure the resolver doesn't choke on this, we'd need to have another operator in the catalog that provides the same API (but I'm pretty sure that's an invariant). We may even want to enforce that provided and required are disjoint sets.

Reply:
That may have been a problem with the old stepwise resolver, but allowing one operator to satisfy its own constraints is natural for the current resolution system. Even thought it means that pointless (tautological) constraints are allowed, they are harmless. A special restriction for this case ends up being extra complexity without a real upside.

Full disclosure: I specifically requested the test case to enshrine this behavior as valid.

Reply:
Yes, this is a testcase that I added just to see how the new resolver would behave in this case as @benluddy mentioned. Since we don't really want to file a separate bug for an additional test I think it's ok to include this test here.

Reply:
The `APISet` is both provided and required on this operator in this catalog as-is I believe